benchmark/benchmark.hpp

```diff
@@ -163,209 +250,164 @@ struct benchmark_interface : public benchmark::Fixture {
     }
 
     virtual std::string name() { return Allocator::name(); }
-    virtual int64_t iterations() { return 10000; }
+
     static void defaultArgs(Benchmark *benchmark) {
         auto *bench =
             static_cast<benchmark_interface<Size, Allocator> *>(benchmark);
-        benchmark->ArgNames(bench->argsName())
-            ->Name(bench->name())
-            ->Iterations(bench->iterations());
+        benchmark->ArgNames(bench->argsName())->Name(bench->name());
     }
-    Size alloc_size;
+
+    std::vector<Size> alloc_sizes;
     Allocator allocator;
 };
 
-// This class benchmarks speed of alloc() operations.
+// This class benchmarks performance of random deallocations and (re)allocations
 template <
     typename Size, typename Alloc,
     typename =
         std::enable_if_t<std::is_base_of<alloc_size_interface, Size>::value>,
     typename =
         std::enable_if_t<std::is_base_of<allocator_interface, Alloc>::value>>
-class alloc_benchmark : public benchmark_interface<Size, Alloc> {
+class multiple_malloc_free_benchmark : public benchmark_interface<Size, Alloc> {
+    using distribution = std::uniform_int_distribution<size_t>;
+    template <class T> using vector2d = std::vector<std::vector<T>>;
+    using base = benchmark_interface<Size, Alloc>;
+
+    int allocsPerIterations = 10;
+    bool thread_local_allocations = true;
+    size_t max_allocs = 0;
+
+    vector2d<alloc_data> allocations;
+    std::vector<unsigned> iters;
+
+    vector2d<next_alloc_data> next;
+    std::vector<std::vector<next_alloc_data>::const_iterator> next_iter;
+    int64_t iterations;
+
   public:
-    size_t max_allocs = 1000;
-    size_t pre_allocs = 0;
     void SetUp(::benchmark::State &state) override {
-        if (state.thread_index() != 0) {
-            return;
-        }
+        auto tid = state.thread_index();
 
-        // unpack arguments
-        int argPos = 0;
-        max_allocs = state.range(argPos++);
-        pre_allocs = state.range(argPos++);
-        // pass rest of the arguments to "alloc_size" and "allocator"
-        argPos = base::alloc_size.SetUp(state, argPos);
-        base::allocator.SetUp(state, argPos);
-
-        // initialize allocations tracking vectors (one per thread)
-        // and iterators for these vectors.
-        allocations.resize(state.threads());
-        iters.resize(state.threads());
-
-        for (auto &i : iters) {
-            i = pre_allocs;
+        if (tid == 0) {
+            // unpack arguments
+            iterations = state.max_iterations;
+            int argPos = 0;
+            max_allocs = state.range(argPos++);
+            thread_local_allocations = state.range(argPos++);
+            base::parseArgs(state, argPos);
+
+            allocations.resize(state.threads());
+            next.resize(state.threads());
+            next_iter.resize(state.threads());
+
+#ifndef WIN32
+            // ensure that system malloc has not memory pooled on heap
+            malloc_trim(0);
```

Comment:
windows: https://stackoverflow.com/questions/72660315/how-to-recycle-memory-of-c-unordered-map-on-windows